<a href="https://colab.research.google.com/github/Muskan2320/Data-Science/blob/main/Dealing_Missing_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Dealing with missing value

There are two approaches to handle the missing values. The one approach is to delete the rows which have the missing values. The second approach is to logically. Filling missing value logically varies in case of categorical and numerical data. In case of categorical data it include it include filling missing value with the category having maximum count. In case of numerical data it include filling missing values using central tendencies like mean, mode or median.

In [1]:
import pandas as pd
import numpy as np

  The 'Pandas' package is a fast, flexible tool used to analyze data. It helps in reading and updating the files like csv files, excel, etc.

  The os module of the Python is normally used to manipulate the installed operating system as well as the PC file system. 

  The 'numpy' package is used to work woth arrays. It can be used for performing mathematical operations and support several other domain like linera algebra, metrices, etc.

  After importing the required packages, now uploading Toyota dataset from the device. Toyota dataset is the data of the cars with the 10 different features specified in the dataset. These features include the fuel type of the car the car i.e. whether it's fuel type is diesel or petrol, similarly other featres like horse power, price of the car, number of doors in the car, etc.

In [2]:
from google.colab import files

upload = files.upload()

Saving Toyota.csv to Toyota.csv


'read_csv' is the command used to read a comma separated file(csv). Simillarly, in Pandas 'read_excel' is used to read excel files. Pandas reads all the blank values 'NaN'. 

Reading csv files sometimes create an 'Unnnamed: 0' as the first column of the dataframe. So, to avoid this we can specify an argument 'index_col= 0' to read_csv() function, then it reads in the first column as the index.

In the Toypta dataset some values are neither blank nor NaN, they may have some special characters like "??", etc. These values are replaced with NaN using an arguement 'na_values = ['list of unwanted values']' to read_csv() function, now pandas reads these unknown values as NaN.

In [3]:
data = pd.read_csv('Toyota.csv',index_col=0,na_values=['??','????'])

data

Price   Age       KM FuelType     HP  MetColor  Automatic    CC  Doors  \
0     13500  23.0  46986.0   Diesel   90.0       1.0          0  2000  three   
1     13750  23.0  72937.0   Diesel   90.0       1.0          0  2000      3   
2     13950  24.0  41711.0   Diesel   90.0       NaN          0  2000      3   
3     14950  26.0  48000.0   Diesel   90.0       0.0          0  2000      3   
4     13750  30.0  38500.0   Diesel   90.0       0.0          0  2000      3   
...     ...   ...      ...      ...    ...       ...        ...   ...    ...   
1431   7500   NaN  20544.0   Petrol   86.0       1.0          0  1300      3   
1432  10845  72.0      NaN   Petrol   86.0       0.0          0  1300      3   
1433   8500   NaN  17016.0   Petrol   86.0       0.0          0  1300      3   
1434   7250  70.0      NaN      NaN   86.0       1.0          0  1300      3   
1435   6950  76.0      1.0   Petrol  110.0       0.0          0  1600      5   

      Weight  
0       1165  
1       1165  
2       1165  
3       1165  
4       1170  
...      ...  
1431    1025  
1432    1015  
1433    1015  
1434    1015  
1435    1114  

[1436 rows x 10 columns]

It is possible that we sometimes copy of data may be needed because it is possible that we don't want to change the original data. 
Copy is of two types:
    1. Shallow Copy
    2. Deep Copy

Shallow copy: It can be done by "shallow_cpy = dataframe_name.copy(deep=False)". By default parameter 'deep' is True. It creates a new object that shares the refrence to original object. Any changes in shallow copy are reflected in dataframe.

Deep Copy: It can be done by "deep_cpy = dataframe_name.copy(deep=True)". It creates a new object copyin the original object. Any changes in deep copy will not reflect in original object.

In [4]:
samp = data.copy(deep=True)

samp

Price   Age       KM FuelType     HP  MetColor  Automatic    CC  Doors  \
0     13500  23.0  46986.0   Diesel   90.0       1.0          0  2000  three   
1     13750  23.0  72937.0   Diesel   90.0       1.0          0  2000      3   
2     13950  24.0  41711.0   Diesel   90.0       NaN          0  2000      3   
3     14950  26.0  48000.0   Diesel   90.0       0.0          0  2000      3   
4     13750  30.0  38500.0   Diesel   90.0       0.0          0  2000      3   
...     ...   ...      ...      ...    ...       ...        ...   ...    ...   
1431   7500   NaN  20544.0   Petrol   86.0       1.0          0  1300      3   
1432  10845  72.0      NaN   Petrol   86.0       0.0          0  1300      3   
1433   8500   NaN  17016.0   Petrol   86.0       0.0          0  1300      3   
1434   7250  70.0      NaN      NaN   86.0       1.0          0  1300      3   
1435   6950  76.0      1.0   Petrol  110.0       0.0          0  1600      5   

      Weight  
0       1165  
1       1165  
2       1165  
3       1165  
4       1170  
...      ...  
1431    1025  
1432    1015  
1433    1015  
1434    1015  
1435    1114  

[1436 rows x 10 columns]

In pandas all the blank values in data are represented by NaN(Not a number). For counting the number of null values in the data we have two functions 'isna' and 'isnull', these functions return the boolean values. 
To count the number of rows with null values in each column, you can use isna().sum() or isnull().sum().

In [5]:
data.isna().sum()  #OR data.isnull().sum()

Price          0
Age          100
KM            15
FuelType     100
HP             6
MetColor     150
Automatic      0
CC             0
Doors          0
Weight         0
dtype: int64

It is possible to find those rows which have one or more columns missing. 'dataframe.isnull()' is to check whether the given value is null or not. 'any(axis=1)' is function representing any column in the row, here axis=1 represents the column(axis=0 represent the row).

In [6]:
missing = data[data.isnull().any(axis=1)]
missing

Price   Age       KM FuelType    HP  MetColor  Automatic    CC Doors  \
2     13950  24.0  41711.0   Diesel  90.0       NaN          0  2000     3   
6     16900  27.0      NaN   Diesel   NaN       NaN          0  2000     3   
7     18600  30.0  75889.0      NaN  90.0       1.0          0  2000     3   
9     12950  23.0  71138.0   Diesel   NaN       NaN          0  1900     3   
15    22000  28.0  18739.0   Petrol   NaN       0.0          0  1800     3   
...     ...   ...      ...      ...   ...       ...        ...   ...   ...   
1428   8450  72.0      NaN   Petrol  86.0       NaN          0  1300     3   
1431   7500   NaN  20544.0   Petrol  86.0       1.0          0  1300     3   
1432  10845  72.0      NaN   Petrol  86.0       0.0          0  1300     3   
1433   8500   NaN  17016.0   Petrol  86.0       0.0          0  1300     3   
1434   7250  70.0      NaN      NaN  86.0       1.0          0  1300     3   

      Weight  
2       1165  
6       1245  
7       1245  
9       1105  
15      1185  
...      ...  
1428    1015  
1431    1025  
1432    1015  
1433    1015  
1434    1015  

[340 rows x 10 columns]

## Descriptive Statistics

'describe' function can be used to get the detailed description of the data.It summarize the central tendencies(mean, median), minimum value, maximum value, distribution, five-number summary excluding the NaN values.

In [7]:
data.describe()

Price          Age             KM           HP     MetColor  \
count   1436.000000  1336.000000    1421.000000  1430.000000  1286.000000   
mean   10730.824513    55.672156   68647.239972   101.478322     0.674961   
std     3626.964585    18.589804   37333.023589    14.768255     0.468572   
min     4350.000000     1.000000       1.000000    69.000000     0.000000   
25%     8450.000000    43.000000   43210.000000    90.000000     0.000000   
50%     9900.000000    60.000000   63634.000000   110.000000     1.000000   
75%    11950.000000    70.000000   87000.000000   110.000000     1.000000   
max    32500.000000    80.000000  243000.000000   192.000000     1.000000   

         Automatic           CC      Weight  
count  1436.000000  1436.000000  1436.00000  
mean      0.055710  1566.827994  1072.45961  
std       0.229441   187.182436    52.64112  
min       0.000000  1300.000000  1000.00000  
25%       0.000000  1400.000000  1040.00000  
50%       0.000000  1600.000000  1070.00000  
75%       0.000000  1600.000000  1085.00000  
max       1.000000  2000.000000  1615.00000

##Approach 1: Dropping those rows which has missing data

We can drop those rows which has only one value NaN or deleting those columns which has NaN values. But dropping these rows or columns will result in loss of lage amount of data. Hence, it is good to logically fill the missing data. Also, it is possible to delete those rows which have all features missing.

In [8]:
## Deleting those rows which has one or more NaN values
data.dropna()  ## OR data.dropna(how='any', axis=0)

Price   Age       KM FuelType     HP  MetColor  Automatic    CC  Doors  \
0     13500  23.0  46986.0   Diesel   90.0       1.0          0  2000  three   
1     13750  23.0  72937.0   Diesel   90.0       1.0          0  2000      3   
3     14950  26.0  48000.0   Diesel   90.0       0.0          0  2000      3   
4     13750  30.0  38500.0   Diesel   90.0       0.0          0  2000      3   
5     12950  32.0  61000.0   Diesel   90.0       0.0          0  2000      3   
...     ...   ...      ...      ...    ...       ...        ...   ...    ...   
1423   7950  80.0  35821.0   Petrol   86.0       0.0          1  1300      3   
1424   7750  73.0  34717.0   Petrol   86.0       0.0          0  1300      3   
1429   8950  78.0  24000.0   Petrol   86.0       1.0          1  1300      5   
1430   8450  80.0  23000.0   Petrol   86.0       0.0          0  1300      3   
1435   6950  76.0      1.0   Petrol  110.0       0.0          0  1600      5   

      Weight  
0       1165  
1       1165  
3       1165  
4       1170  
5       1170  
...      ...  
1423    1015  
1424    1015  
1429    1065  
1430    1015  
1435    1114  

[1096 rows x 10 columns]

You can see the that initially our data had 4096 rows but after removing rows with NaN values data has only 1096 rows left.

In [9]:
## Delete only those rows which has all values as NaN
data.dropna(how = 'all')

Price   Age       KM FuelType     HP  MetColor  Automatic    CC  Doors  \
0     13500  23.0  46986.0   Diesel   90.0       1.0          0  2000  three   
1     13750  23.0  72937.0   Diesel   90.0       1.0          0  2000      3   
2     13950  24.0  41711.0   Diesel   90.0       NaN          0  2000      3   
3     14950  26.0  48000.0   Diesel   90.0       0.0          0  2000      3   
4     13750  30.0  38500.0   Diesel   90.0       0.0          0  2000      3   
...     ...   ...      ...      ...    ...       ...        ...   ...    ...   
1431   7500   NaN  20544.0   Petrol   86.0       1.0          0  1300      3   
1432  10845  72.0      NaN   Petrol   86.0       0.0          0  1300      3   
1433   8500   NaN  17016.0   Petrol   86.0       0.0          0  1300      3   
1434   7250  70.0      NaN      NaN   86.0       1.0          0  1300      3   
1435   6950  76.0      1.0   Petrol  110.0       0.0          0  1600      5   

      Weight  
0       1165  
1       1165  
2       1165  
3       1165  
4       1170  
...      ...  
1431    1025  
1432    1015  
1433    1015  
1434    1015  
1435    1114  

[1436 rows x 10 columns]

In [10]:
## Dropping those columns which has NaN values
data.dropna(axis = 1)

Price  Automatic    CC  Doors  Weight
0     13500          0  2000  three    1165
1     13750          0  2000      3    1165
2     13950          0  2000      3    1165
3     14950          0  2000      3    1165
4     13750          0  2000      3    1170
...     ...        ...   ...    ...     ...
1431   7500          0  1300      3    1025
1432  10845          0  1300      3    1015
1433   8500          0  1300      3    1015
1434   7250          0  1300      3    1015
1435   6950          0  1600      5    1114

[1436 rows x 5 columns]

##Approach 2: Filling missing values logically

####Fill NaN values for numerical features
For numerical features the NaN values can be filled with mean of the sample or median of the sample.

In [11]:
## Mean and median of Age feature don't have much difference so we can replace NaN values with mean
samp['Age'].fillna(samp['Age'].mean(), inplace=True)    ## Filling NaN values with mean

## Mean and median of KM feature have a huge difference so we can't replace NaN values with mean. 
## The value of mean is very large as compared to median it may be because of any outliers present in KM. 
samp['KM'].fillna(samp['KM'].median(), inplace=True)    ## Filling NaN values with median

## Mean and median of HP feature don't have much difference so we can replace NaN values with mean
samp['HP'].fillna(samp['HP'].mean(), inplace=True)

Get the number of null values in each column

In [12]:
samp.isnull().sum()

Price          0
Age            0
KM             0
FuelType     100
HP             0
MetColor     150
Automatic      0
CC             0
Doors          0
Weight         0
dtype: int64

####Fill NaN values for categorical features
For categorical features the NaN values can be filled with the category which has the maximum count.

In [13]:
samp['FuelType'].fillna(samp['FuelType'].value_counts().index[0], inplace=True)
samp['MetColor'].fillna(samp['MetColor'].value_counts().index[0], inplace=True)

Get the number of null values in each column

In [14]:
samp.isnull().sum()

Price        0
Age          0
KM           0
FuelType     0
HP           0
MetColor     0
Automatic    0
CC           0
Doors        0
Weight       0
dtype: int64